libr/core/linux_heap_glibc.c

```diff
@@ -849,56 +853,114 @@ void GH(print_heap_fastbin)(RCore *core, GHT m_arena, MallocState *main_arena, G
+	GHT tcache_tmp = GHT_MAX;
+	RConsPrintablePalette *pal = &r_cons_singleton ()->context->pal;
+	size_t i;
+	for (i = 0; i < TCACHE_MAX_BINS; i++) {
+		if (tcache_heap->counts[i] > 0) {
+			PRINT_GA ("bin :");
+			PRINTF_BA ("%2d",i);
+			PRINT_GA (", items :");
+			PRINTF_BA ("%2d",tcache_heap->counts[i]);
+			PRINT_GA (", fd :");
+
+			PRINTF_BA ("0x%"PFMT64x, (GHT)(tcache_heap->entries[i]) - GH(HDR_SZ));
+			if (tcache_heap->counts[i] > 1) {
+				tcache_fd = (GHT) tcache_heap->entries[i];
+				size_t n;
+				for (n = 1; n < tcache_heap->counts[i]; n++) {
+					bool r = r_io_read_at (core->io, tcache_fd, (ut8 *)&tcache_tmp, sizeof (GHT));
+					if (r) {
+						tcache_tmp = r_read_le (&tcache_tmp);
+						PRINTF_BA ("->0x%"PFMT64x, tcache_tmp - TC_HDR_SZ);
+						tcache_fd = tcache_tmp;
+					} else {
+						break;
+					}
+				}
+			}
+			PRINT_BA ("\n");
+		}
+	}
+}
+
+static void GH(parse_tcache_old)(RCore *core, GH(RHeapTcacheOld) *tcache_heap) {
+	GHT tcache_fd = GHT_MAX;
+	GHT tcache_tmp = GHT_MAX;
+	RConsPrintablePalette *pal = &r_cons_singleton ()->context->pal;
+	size_t i;
+	for (i = 0; i < TCACHE_MAX_BINS; i++) {
+		if (tcache_heap->counts[i] > 0) {
+			PRINT_GA ("bin :");
+			PRINTF_BA ("%2d",i);
+			PRINT_GA (", items :");
+			PRINTF_BA ("%2d",tcache_heap->counts[i]);
+			PRINT_GA (", fd :");
+
+			PRINTF_BA ("0x%"PFMT64x, (GHT)(tcache_heap->entries[i]) - GH(HDR_SZ));
+			if (tcache_heap->counts[i] > 1) {
+				tcache_fd = (GHT) tcache_heap->entries[i];
+				size_t n;
+				for (n = 1; n < tcache_heap->counts[i]; n++) {
+					bool r = r_io_read_at (core->io, tcache_fd, (ut8 *)&tcache_tmp, sizeof (GHT));
+					if (r) {
+						tcache_tmp = r_read_le (&tcache_tmp);
+						PRINTF_BA ("->0x%"PFMT64x, tcache_tmp - TC_HDR_SZ);
+						tcache_fd = tcache_tmp;
+					} else {
+						break;
+					}
+				}
+			}
+			PRINT_BA ("\n");
+		}
 	}
+}
+
+static void GH(print_tcache_instance)(RCore *core, GHT m_arena, MallocState *main_arena) {
+	r_return_if_fail (core && core->dbg && core->dbg->maps);
+
 	const int tcache = r_config_get_i (core->config, "dbg.glibc.tcache");
 	if (!tcache) {
 		return;
 	}
-	GHT brk_start = GHT_MAX, brk_end = GHT_MAX, tcache_fd = GHT_MAX, initial_brk = GHT_MAX ;
+	GHT brk_start = GHT_MAX, brk_end = GHT_MAX, initial_brk = GHT_MAX ;
 	GH(get_brks) (core, &brk_start, &brk_end);
-	GHT tcache_tmp = GHT_MAX, tcache_start = GHT_MAX;
+	GHT tcache_start = GHT_MAX;
 	RConsPrintablePalette *pal = &r_cons_singleton ()->context->pal;
 
-	tcache_start = ((brk_start >> 12) << 12) + GH(HDR_SZ);
+	tcache_start = brk_start + 0x10;
 	GHT fc_offset = GH(tcache_chunk_size) (core, brk_start);
 	initial_brk = brk_start + fc_offset;
 	if (brk_start == GHT_MAX || brk_end == GHT_MAX || initial_brk == GHT_MAX) {
 		eprintf ("No heap section\n");
 		return;
 	}
 
-	GH(RHeapTcache) *tcache_heap = R_NEW0 (GH(RHeapTcache));
-	if (!tcache_heap) {
-		return;
+	double glibc_version = core->dbg->glibc_version;
+	GH(RHeapTcache) *tcache_heap;
```

Comment:
Mh.. I still think having an union would have been easier to implement/read :( it would have avoided having 2 variables to deal with.